_drafts/2020-09-17-fixing-an-n1-performance-problem-in-xaf-xpo-with-totally-undocumented-apis.md

```diff
@@ -0,0 +1,948 @@
+To tell XPO about our database, we use attributes to specify relationships between entities. That allows XPO to *guess* what SQL statements it should generate. This is a very powerful abstraction, because you don't need to constantly think about SQL and can focus on business logic. That's fine for the most part but if the number of records grow (or tables and relationships get more complicated) that guess can go horribly wrong.
+
+Expensive queries are also something that can occur with ORMs (like massive JOIN's) but that's not the focus of this blog post.
+
+I'll give you a litte example in **very** naive C#.
+
+> PLEASE NEVER DO SOMETHING LIKE THIS IN PRODUCTION. YOU HAVE BEEN WARNED. (Or I insist you NEED to [book some consulting hour's](https://www.delegate.at) from me)
+
+```cs
+[DefaultClassOptions]
+[DefaultProperty(nameof(HourSum))]
+public class SlowOffer : BaseObject
+{
+    public SlowOffer(Session session) : base(session) { }
+
+    private string _Name;
+    [Persistent("Name")]
+    public string Name { get => _Name; set => SetPropertyValue(nameof(Name), ref _Name, value); }
+
+    [NonPersistent]
+    public int HourSum
+    {
+        get
+        {
+            var sum = 0;
+            foreach (var item in OfferItems)
+            {
+                sum += item.Hours;
+            }
+            return sum;
+        }
+    }
+
+    [Association, Aggregated]
+    public XPCollection<SlowOfferItem> OfferItems => GetCollection<SlowOfferItem>(nameof(OfferItems));
+}
+
+public class SlowOfferItem : BaseObject
+{
+    public SlowOfferItem(Session session) : base(session) { }
+
+    private int _Hours;
+    [Persistent("Hours")]
+    public int Hours { get => _Hours; set => SetPropertyValue(nameof(Hours), ref _Hours, value); }
+
+    private SlowOffer _SlowOffer;
+    [Persistent("SlowOffer"), Association]
+    public SlowOffer SlowOffer { get => _SlowOffer; set => SetPropertyValue(nameof(SlowOffer), ref _SlowOffer, value); }
+}
+
+```
+
+> For all examples I use 300 `Offers` with 1000 `OfferItems` each. It's sorted by the `HourSum` property acending. I use SqlServer LocalDb for this tests on a Intel(R) Core(TM) i7-4770K CPU @ 3.50GHz with 16 GB of memory and a two SSD disks Raid 0 setup.  
+The build configuration is set on Debug with an attached debugger.  
+XPO log verbosity = 4  
+XAF log verbosity = 3  
+This means those numbers *will increase in Release* configuration, so we setting the base line pessimistic (and also aim for an better developer feedback cycle).
```

Comment:
```suggestion
This means those numbers *will improve in Release* configuration, so we set a pessimistic base line (and also aim for a better developer feedback cycle).
```

I'm guessing you mean that the timed values will decrease in RELEASE, so the 'numbers' will improve.